numpy/core/src/umath/dispatching.c

```diff
@@ -1240,3 +1240,40 @@ install_logical_ufunc_promoter(PyObject *ufunc)
 
     return PyUFunc_AddLoop((PyUFuncObject *)ufunc, info, 0);
 }
+
+/*
+ * Return the PyArrayMethodObject or PyCapsule that matches a registered
+ * tuple of identical dtypes. Return a borrowed ref of the first match.
+ */
+NPY_NO_EXPORT PyObject *
+get_info_no_cast(PyUFuncObject *ufunc, PyArray_DTypeMeta *op_dtype,
+                 int ndtypes)
+{
+    PyObject *t_dtypes = PyTuple_New(ndtypes);
+    if (t_dtypes == NULL) {
+        return NULL;
+    }
+    for (int i=0; i < ndtypes; i++) {
+        PyTuple_SetItem(t_dtypes, i, (PyObject *)op_dtype);
+    }
+    PyObject *loops = ufunc->_loops;
+    Py_ssize_t length = PyList_Size(loops);
+    for (Py_ssize_t i = 0; i < length; i++) {
+        PyObject *item = PyList_GetItem(loops, i);
+        PyObject *cur_DType_tuple = PyTuple_GetItem(item, 0);
+        int cmp = PyObject_RichCompareBool(cur_DType_tuple,
+                                           t_dtypes, Py_EQ);
+        if (cmp < 0) {
+            Py_DECREF(t_dtypes);
+            return NULL;
+        }
+        if (cmp == 0) {
+            continue;
+        }
+        /* Got the match */
+        Py_DECREF(t_dtypes);
+        return PyTuple_GetItem(item, 1);
+    }
+    Py_DECREF(t_dtypes);
+    Py_RETURN_NONE;
+}
```

Comment:
Should I add a test for these error paths?

Reply:
Not for my sake, I don't really believe in unit-testing when we have integration tests and the unit test is going to be tedious.  If we ever move this into public API, maybe.

I have to do a careful path, but really the only thing for me remains that I am not sure I like the API of just gluing on the loop much, and as I said, I am willing to live with it since it is replaceable easy enough.

Reply:
> I am not sure I like the API of just gluing on the loop much

I didn't want to touch the already-overloaded `PyUFunc_FromFuncAndDataAndSignatureAndIdentity`. We might want to create a  `PyUfunc_FromSpec` or so.

Reply:
Sorry, definitely not touching that :).  I am still just a bit squirmish about the new loop signature and where to put it exactly.